src/core/net/dns_types.cpp

```diff
@@ -626,30 +631,56 @@ Error Name::LabelIterator::AppendLabel(Message &aMessage) const
 
 bool Name::IsSubDomainOf(const char *aName, const char *aDomain)
 {
-    bool     match        = false;
-    uint16_t nameLength   = StringLength(aName, kMaxNameLength);
-    uint16_t domainLength = StringLength(aDomain, kMaxNameLength);
+    bool     match             = false;
+    bool     nameEndsWithDot   = false;
+    bool     domainEndsWithDot = false;
+    uint16_t nameLength        = StringLength(aName, kMaxNameLength);
+    uint16_t domainLength      = StringLength(aDomain, kMaxNameLength);
 
     if (nameLength > 0 && aName[nameLength - 1] == kLabelSeperatorChar)
     {
+        nameEndsWithDot = true;
         --nameLength;
     }
 
     if (domainLength > 0 && aDomain[domainLength - 1] == kLabelSeperatorChar)
     {
+        domainEndsWithDot = true;
         --domainLength;
     }
 
     VerifyOrExit(nameLength >= domainLength);
+
     aName += nameLength - domainLength;
 
     if (nameLength > domainLength)
     {
         VerifyOrExit(aName[-1] == kLabelSeperatorChar);
     }
-    VerifyOrExit(memcmp(aName, aDomain, domainLength) == 0);
 
-    match = true;
+    if (nameEndsWithDot == domainEndsWithDot)
+    {
+        // Both `aName` and `aDomain` end with dot or both do not.
+        // We can compare the strings directly.
+
+        match = StringMatch(aName, aDomain, kStringCaseInsensitiveMatch);
+    }
+    else if (nameEndsWithDot)
+    {
+        // `aName` ends with dot, but `aDomain` does not.
+        // We check that `aName` contains `aDomain` as a
+        // prefix string.
+
+        match = StringStartsWith(aName, aDomain, kStringCaseInsensitiveMatch);
+    }
+    else
+    {
+        // `aDomain` ends with dot, but `aName` does not.
+        // We use check that `aName` contains `aDomain` as
+        // a prefix string.
+
+        match = StringStartsWith(aDomain, aName, kStringCaseInsensitiveMatch);
+    }
```

Comment:
nit: The code is not very intuitive.

Reply:
I added the comments for that reason.

FWIW, the `IsSubDomainOf()` allows either `aName` or `aDomain` to include or exclude the last `.` character.  I personally preferred a bit better if we required both to include the `.`. This is what adds the complexity.

Reply:
I combined the comments into one at the top of `if()` block in the new push.